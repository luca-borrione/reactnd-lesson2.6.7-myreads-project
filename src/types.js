
import PropTypes from 'prop-types'

/**
 * @typedef {Object} TBook
 * @description
 * Custom type to define a book internally.
 * It's filled with properties coming from the
 * book literal object fetched using the BooksAPI
 * @property {string[]} [authors] - List of the authors names
 * @property {string} id - The book unique id
 * @property {string} [shelf] - The shelf unique name
 * @property {string} [thumbnail] - URL for the cover preview image
 * @property {string} title - The book's title
 */

export const TBook = {
	authors: PropTypes.arrayOf(PropTypes.string.isRequired),
	id: PropTypes.string.isRequired,
	shelf: PropTypes.string,
	thumbnail: PropTypes.string,
	title: PropTypes.string.isRequired
};

