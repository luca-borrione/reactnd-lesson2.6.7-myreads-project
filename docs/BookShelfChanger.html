<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>BookShelfChanger - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a><ul class='methods'><li data-type='method'><a href="App.html#fetchAllBooks">fetchAllBooks</a></li><li data-type='method'><a href="App.html#getBookShelf">getBookShelf</a></li><li data-type='method'><a href="App.html#render">render</a></li><li data-type='method'><a href="App.html#updateBookShelf">updateBookShelf</a></li></ul></li><li><a href="Book.html">Book</a><ul class='methods'><li data-type='method'><a href="Book.html#onShelfChange">onShelfChange</a></li><li data-type='method'><a href="Book.html#render">render</a></li></ul></li><li><a href="BooksGrid.html">BooksGrid</a><ul class='methods'><li data-type='method'><a href="BooksGrid.html#render">render</a></li></ul></li><li><a href="BookShelf.html">BookShelf</a><ul class='methods'><li data-type='method'><a href="BookShelf.html#render">render</a></li></ul></li><li><a href="BookShelfChanger.html">BookShelfChanger</a><ul class='methods'><li data-type='method'><a href="BookShelfChanger.html#render">render</a></li><li data-type='method'><a href="BookShelfChanger.html#selectionChanged">selectionChanged</a></li></ul></li><li><a href="BooksList.html">BooksList</a><ul class='methods'><li data-type='method'><a href="BooksList.html#render">render</a></li></ul></li><li><a href="SearchPage.html">SearchPage</a><ul class='methods'><li data-type='method'><a href="SearchPage.html#render">render</a></li></ul></li></ul><h3>Modules</h3><ul><li><a href="module-Constants.html">Constants</a></li></ul><h3>Namespaces</h3><ul><li><a href="StringUtils.html">StringUtils</a><ul class='methods'><li data-type='method'><a href="StringUtils.html#.makeTitle">makeTitle</a></li><li data-type='method'><a href="StringUtils.html#.ucFirst">ucFirst</a></li><li data-type='method'><a href="StringUtils.html#.ucWords">ucWords</a></li></ul></li></ul><h3>Global</h3><ul><li><a href="global.html#TShelfKey">TShelfKey</a></li></ul>
</nav>

<div id="main">
    
    <h1 class="page-title">BookShelfChanger</h1>
    

    




<section>

<header>
    
        <h2>
        BookShelfChanger
        </h2>
        
            <div class="class-description"><p>Shows a selection to change the shelf of the book</p></div>
        
    
</header>

<article>
    

    
        <h3 class="subsection-title">Extends</h3>

        


    <ul>
        <li>React.Component</li>
    </ul>


    

    

    

     

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id="propTypes"><span class="type-signature"></span>propTypes<span class="type-signature"></span></h4>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>



    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>propTypes</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            

            

            <td class="description last"><p>Intended types passed to the component</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>shelf</code></td>
            

            <td class="type">
            
                
<span class="param-type"><a href="global.html#TShelfKey">TShelfKey</a></span>


            
            </td>

            

            

            <td class="description last"><p>The shelf key currently associated with the book</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>propTypes.onShelfChange</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last"><p>method of the parent <a href="Book.html">Book</a> component which handles a new shelf value</p></td>
        </tr>

    
    </tbody>
</table>













        
            
<h4 class="name" id="state"><span class="type-signature">(private) </span>state<span class="type-signature"></span></h4>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>



    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>state</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            

            

            <td class="description last">
                <h6>Properties</h6>

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>selectedValue</code></td>
            

            <td class="type">
            
                
<span class="param-type"><a href="global.html#TShelfKey">TShelfKey</a></span>


            
            </td>

            

            

            <td class="description last"><p>Shelf key associated with the current selection</p></td>
        </tr>

    
    </tbody>
</table>

            </td>
        </tr>

    
    </tbody>
</table>













        
    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    <h4 class="name" id="render"><span class="type-signature"></span>render<span class="signature">()</span><span class="type-signature"> &rarr; {ReactElement}</span></h4>

    




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>





<div class="description">
    <p>Creates the selection containing all the available shelves.</p>
</div>























<h5>Returns:</h5>

        


<dl class="param-type">
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">ReactElement</span>


    </dd>
</dl>

    


        
            

    

    <h4 class="name" id="selectionChanged"><span class="type-signature"></span>selectionChanged<span class="signature">(event)</span><span class="type-signature"> &rarr; {void}</span></h4>

    




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>





<div class="description">
    <p>Handles the change event of the user selection.<br>
Firstly it updates the internal state, then it triggers the onShelfChange method
of the parent <a href="Book.html">Book</a> component to handle the new shelf value.</p>
</div>











    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>event</code></td>
            

            <td class="type">
            
                
<span class="param-type">SyntheticEvent</span>


            
            </td>

            

            

            <td class="description last"></td>
        </tr>

    
    </tbody>
</table>














<h5>Returns:</h5>

        


<dl class="param-type">
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    


        
    

    

    
</article>

</section>




    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Thu Nov 15 2018 23:47:23 GMT+0000 (GMT) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>


</body>
</html>