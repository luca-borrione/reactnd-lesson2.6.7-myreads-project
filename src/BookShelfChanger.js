import React from 'react';
import PropTypes from 'prop-types';
import { SHELF_TITLE } from './Constants';
import { TBook } from './types';

/**
 * @class BookShelfChanger
 * @extends React.Component
 * @classdesc
 * Shows a selection to change the shelf of the book
 * @hideconstructor
 */
class BookShelfChanger extends React.Component {

	/**
	 * @property {Object} propTypes - Intended types passed to the component
	 * @property {TShelfKey} shelf - The shelf key currently associated with the book
	 * @property {function} propTypes.onShelfChange - method of the parent [Book]{@link Book} component which handles a new shelf value
	 * @static
	 */
	static propTypes = {
		book: PropTypes.shape(TBook).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	constructor(props) {
		super(props);
		this.onSelectChange = this.onSelectChange.bind(this);
	}


	/**
	 * @property {Object} state
	 * @property {TShelfKey} state.selectedValue - Shelf key associated with the current selection
	 * @private
	 */
	state = {
		selectedValue: ''
	};


	// The shelf key coming from the props becomes stored internally as state
	static getDerivedStateFromProps(nextProps, prevState) {
		if (prevState.selectedValue === '') {
			return {
				selectedValue: nextProps.book.shelf
			};
		}
		return null;
	}


	/**
	 * @description
	 * Handles the change event of the user selection.<br>
	 * Firstly it updates the internal state, then it triggers the
	 * [updateBookShelf]{@link App#updateBookShelf} method of the App component
	 * to update the list of books in the shelves
	 * @param {SyntheticEvent} event
	 * @returns {void}
	 */
	onSelectChange(event) {
		const { book, updateBookShelf } = this.props;

		this.setState({
			selectedValue: event.target.value
		}, () => {
			updateBookShelf(book, this.state.selectedValue);
		});
	}


	/**
	 * @description
	 * Creates the selection containing all the available shelves.
	 * @returns {ReactElement}
	 */
	render() {
		return (
			<select value={this.state.selectedValue} onChange={this.onSelectChange}>
				<option value="move" disabled>Move to...</option>
				{Object.entries(SHELF_TITLE).map( ([ shelfKey, shelfTitle ], index) => (
					<option key={index} value={shelfKey}>{shelfTitle}</option>
				))}
			</select>
		);
	}
}


export default BookShelfChanger;
