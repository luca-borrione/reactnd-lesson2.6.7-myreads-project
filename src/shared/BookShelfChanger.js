import React from 'react';
import PropTypes from 'prop-types';
import { SHELF_TITLE } from './constants';
import { TBook } from './types';


/**
 * @module
 * @name BookShelfChanger
 * @extends React.Component
 * @description
 * Shows a selection that allows the user to change the shelf of the book
 */
class BookShelfChanger extends React.Component {

	/**
	 * @member
	 * @name STATUS
	 * @description Collection of possible status
	 * @property {string} BUSY - The user selectd a different shelf for this book and the App is busy changing the shelf remotely
	 * @property {string} READY - Either the component received the current shelf in the props or it successfully changed it remotely
	 * @static
	 */
	static STATUS = {
		BUSY: 'STATUS.BUSY',
		READY: 'STATUS.READY'
	};
	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook} book - The book the changer is associated with
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		book: PropTypes.shape(TBook).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	constructor(props) {
		super(props);
		this.onSelectChange = this.onSelectChange.bind(this);
	}


	/**
	 * @member
	 * @name state
	 * @property {TShelfKey} selectedValue - the current selected shelf key
	 * @property {STATUS} status - one of the possible [STATUS]{@link module:BookShelfChanger.STATUS}
	 * @private
	 */
	state = {
		selectedValue: '',
		status: ''
	};


	// The shelf key coming from the props becomes stored internally as state
	static getDerivedStateFromProps(nextProps, prevState) {
		if (prevState.selectedValue === '') {
			return {
				selectedValue: nextProps.book.shelf,
				status: BookShelfChanger.STATUS.READY
			};
		}
		return null;
	}


	/**
	 * @method
	 * @description
	 * Handles the change event of the user selection.
	 * Firstly it updates the internal state, then it triggers the
	 * [updateBookShelf]{@link module:App~updateBookShelf} method of the App component
	 * to update the list of books in the shelves
	 * @param {SyntheticEvent} event
	 * @returns {void}
	 * @private
	 */
	onSelectChange(event) {
		const { STATUS } = this.constructor;
		const { book, updateBookShelf } = this.props;

		this.setState({
			selectedValue: event.target.value,
			status: STATUS.BUSY
		}, async () => {
			await updateBookShelf(book, this.state.selectedValue);

			// - Moving shelf in MySearch will cause the book to unmount,
			// in this case the new shelf will be passed throught the props.
			// - Moving shelf in the SearchPage won't cause the book to be umounted,
			// but the changer won't be renderer again either, so we need
			// to set change the status to trigger a re-render.
			// We need to check if the component is mounted to prevent triggering
			// an error in the console
			if (this._isMounted) {
				this.setState({
					status: STATUS.READY
				});
			}
		});
	}


	componentDidMount() {
		this._isMounted = true;
	}


	componentWillUnmount() {
		this._isMounted = false;
	}


	/**
	 * @method
	 * @name render
	 * @description
	 * Creates the selection containing all the available shelves.
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { STATUS } = this.constructor;
		const { status } = this.state;

		return (
			<div className="book-shelf-changer">
				{status === STATUS.BUSY && (
					<div id="loading"></div>
				)}
				<select value={this.state.selectedValue} onChange={this.onSelectChange}>
					<option value="move" disabled>Move to...</option>
					{Object.entries(SHELF_TITLE).map( ([ shelfKey, shelfTitle ], index) => (
						<option key={index} value={shelfKey}>{shelfTitle}</option>
					))}
				</select>
			</div>
		);
	}
}


export default BookShelfChanger;
