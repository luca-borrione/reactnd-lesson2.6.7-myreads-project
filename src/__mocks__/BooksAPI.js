import * as data from './allBooks.json'
import { TShelfKey } from '../shared/types'

const DELAY = 10;

const getDelay = () => DELAY * (Math.floor(Math.random() * 10) + 1); // random between 1 - 10

export const MOCKED = true;

let allBooks;
export const __resetBooks = () => {
	allBooks = JSON.parse(JSON.stringify(data.books)); // cloning
};
__resetBooks();

const getBooksInShelves = () =>
	allBooks.filter( ({ shelf }) => shelf !== TShelfKey.NONE );


export const getAll = () => {
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(booksInShelves);
		}, delay);
	});
};

export const get = bookId => {
	const book = allBooks.find( ({ id }) => id === bookId );
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(book);
		}, delay);
	});
};

const formatUpdatedBooks = (books) => {
	const fBooks = {};

	Object.keys(TShelfKey).forEach( key => {
		const shelfKey = TShelfKey[key];
		if (shelfKey !== TShelfKey.NONE) {
			fBooks[shelfKey] = books
							.filter( ({ shelf }) => shelf === shelfKey )
							.map( book => book.id );
		}
	});

	return fBooks
};

export const update = (book, shelf) => {
	const existingBook = allBooks.find( ({ id }) => id === book.id );
	existingBook.shelf = shelf;
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(formatUpdatedBooks(booksInShelves));
		}, delay);
	});
};

export const search = (query, maxResults) => {
	const re = new RegExp(`^${query}`, 'i');
	const result = allBooks.filter( ({ title, authors }) => {
		return re.test(title);
	})
	.slice(0, maxResults);

	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(result);
		}, delay);
	});
};

