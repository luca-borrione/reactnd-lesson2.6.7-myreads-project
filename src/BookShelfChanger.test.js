import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import BookShelfChanger from './BookShelfChanger';
import * as BooksAPI from './BooksAPI'; // mocked
import { TShelfKey } from './types';

describe('BookShelfChanger', () => {

	let book;

	const updateBookShelf = () => {};

	beforeEach( async () => {
		if (!book) {
			book = await BooksAPI.get('nggnmAEACAAJ');
		}
	});


	it('renders without crashing', () => {
		const props = {
			book,
			updateBookShelf
		};
		const div = document.createElement('div');
		ReactDOM.render(<BookShelfChanger {...props} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});


	it('renders correctly', () => {
		const props = {
			book,
			updateBookShelf
		};
		const tree = TestRenderer
			.create(<BookShelfChanger {...props} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});


	describe('state behaviour', () => {

		it('changes the selectedValue state when a new shelf is selected and triggers the prop updateBookShelf method', () => {
			const props = {
				book,
				updateBookShelf: jest.fn()
			};

			const initialShelf = book.shelf;
			const shelf = TShelfKey.WANT_TO_READ;

			const wrapper = mount(<BookShelfChanger {...props} />);
			const component = wrapper.instance();
			expect(component.state.selectedValue).toBe(initialShelf);

			const select = wrapper.find('select');
			select.simulate('change', {
				target: { value : shelf}
			});
			expect(component.state.selectedValue).toBe(shelf);

			expect(props.updateBookShelf).toHaveBeenCalledTimes(1);
			expect(props.updateBookShelf).toHaveBeenCalledWith(props.book, shelf);
		});

	});

});
