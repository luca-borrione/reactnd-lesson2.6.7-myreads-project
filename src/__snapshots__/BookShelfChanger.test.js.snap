// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookShelfChanger renders correctly 1`] = `
<select
  onChange={[Function]}
  value="currentlyReading"
>
  <option
    disabled={true}
    value="move"
  >
    Move to...
  </option>
  <option
    value="currentlyReading"
  >
    Currently Reading
  </option>
  <option
    value="wantToRead"
  >
    Want to Read
  </option>
  <option
    value="read"
  >
    Read
  </option>
  <option
    value="none"
  >
    None
  </option>
</select>
`;
